exasol/ai/mcp/server/main.py

```diff
@@ -238,49 +252,133 @@ def create_mcp_server(
+    if (ENV_USER not in env) and not (use_open_id and (ENV_USERNAME_CLAIM in env)):
+        raise ValueError(
+            "The inferred authentication method requires a database username"
+        )
+    return common_kwargs, use_open_id
+
+
+def _create_connection_pool(env: dict[str:Any]) -> NamedObjectPool[ExaConnection]:
+    pool_size = int(env.get(ENV_POOL_SIZE, DEFAULT_CONN_POOL_SIZE))
+    return NamedObjectPool(capacity=pool_size, cleanup=lambda conn: conn.close())
+
+
+def get_connection_factory(
+    env: dict[str:Any], **extra_kwargs
+) -> Callable[[], ContextManager[ExaConnection]]:
+    """
+    Returns the pyexasol connection factory required by a DBConnection object.
+    Authentication method will be inferred from the provided configuration
+    parameters. Currently, the parameters come from environment variables.
+    Going forward, the configuration parameters will be kept in the NBC secret store.
+
+    The MCP server supports the same authentication methods as pyexasol. Currently,
+    these are password and an OpenID token.
+
+    The MCP server can be deployed in two ways: locally or as a remote http server.
+    In the latter case the server works in the multiple user mode and its tools must be
+    protected with OAuth2 authorization. The server can identify the user by looking
+    at the claims in the access token. Most identity providers allow setting a custom
+    claim or offer a choice of standard claims that can be used to store the DB
+    username. The server needs to know the name of this claim.
+
+    This gives us three basic options for the database connection:
+    - The server is configured to use a specific username and either password or bearer
+      token authentication. This works for both single and multiple user modes.
+      The server's DB user must have the permission that is the least common denominator
+      of the permissions of the users that are allowed to access the MCP server.
+
+    - The server extracts the DB username, along with the token, from the MCP Auth
+      context and uses that to open the connection. This option is suitable for
+      multiuser mode, when the following two conditions are met:
+      1. The users' authentication with the chosen identity provider is configured to
+         add their DB usernames as a claim in the access token.
+      2. The correspondent DB users are also authenticated using OpenID, with access
+         token (refresh token is currently not supported). The database verifies the
+         token with the same identity provider as the MCP server.
+
+    - The server extracts the access token from the MCP Auth context, but uses its
+      own username. This is a workaround for the multiuser mode, when either of the
+      above conditions cannot be met. For this option we will not cache the connection,
+      otherwise a token issued for one user could be used to execute a query requested
```

Comment:
Maybe add a link to the underlying problem here:
https://en.wikipedia.org/wiki/Confused_deputy_problem

Reply:
I think a link to confuse deputy problem will only confuse the reader. Sorry for the pan. In the actual fact other user's tokens will technically work. The DB only checks if the token is good or not (i.e. can be decrypted). Any good token is as good as any other. I only disallow this to avoid cognitive dissonance.